### ISSUE TITLE
Simultaneous inference with different input shapes

### ISSUE THREAD

**Issue by JustasZJ** [openvinotoolkit/openvino#6552]
**TL;DR: I am looking for a fast way to simultaneously execute networks with the same topology, but different input shapes on different cores, without using too much memory.**

I am trying to use openvino for inference on different cores as follows:

1. Construct network topology from scratch using ngraph, save the parameter and the result vector.
...
2. Do inference on different cores using `inferRequest.Infer()`.

In-between stages 1 and 2 I need some logic, which obtains an executable network with the necessary shape for inference on the current core. For this I have tried two methods.

**First method**
When a core wants to do inference:
1. Create `std::shared_ptr<ngraph::Function>` using the saved parameter and results.
2. Wrap the function into a `CNNNetwork`.
3. Reshape the `CNNNetwork` with the required shape.
4. Update the saved parameter with the reshaped one (without this, there would be trouble loading the network on the next core).
5. Create `InferenceEngine::ExecutableNetwork` from the `CNNNetwork`.

The problem with this method, is that I need to put everything under a mutex, which causes a very significant slowdown, otherwise I get errors. This seems to be related to [https://docs.openvinotoolkit.org/2021.1/openvino_docs_IE_DG_ShapeInference.html](https://docs.openvinotoolkit.org/2021.1/openvino_docs_IE_DG_ShapeInference.html) `Do not use runtime reshaping methods simultaneously, especially do not call the CNNNetwork::reshape method after you use InferenceEngine::CNNNetwork::setBatchSize.`

**Second method**
Only once create `std::shared_ptr<ngraph::Function>` using the saved parameter and results and wrap it into `CNNNetwork`, then save it into a variable `savedNetwork`.
When a core wants to do inference:
1. `CNNNetwork network = savedNetwork`.
2. Reshape `network`.
3. Create `InferenceEngine::ExecutableNetwork` from `network`.

Again, this only works if stages 1-3 are under one mutex, which causes a very significant slowdown. From what I understand, the issue here is because both of the copies manage the same ngraph function.

The only real solution that I can think of is to create multiple different ngraph functions and use a different function on each core, but this seems very tedious and would most likely use a lot of memory. 

Are there any other (smoother) ways of working around my issue? I would be very thankful for your suggestion.

Note: I am using openvino 2021.2, but looking to upgrade to the 2021.4 LTS soon.

**Comment by ilya-lavrenov** [comment]
@ilyachur if we create multiple copies of ngraph function, they share the constants, right? In this case approach 2 should work

@JustasZJ actually I wonder why do we need to put everything under mutex.. once the executable network is compiled, ngraph::Function can be used for different reasons like reshape and compile a executable network. Could you please provide a small repro for the scenario you are implementing?

**Comment by JustasZJ** [comment]
@ilya-lavrenov Thank you for the follow-up. I am sorry, but I cannot provide a full repro. In essence, the part which caused trouble looked like this: [https://pastebin.com/a4GaLnJ0](https://pastebin.com/a4GaLnJ0).

Just to clarify from my earlier message, the variable `savedNetwork` is shared between all cores, and not created for each core separately. Also, `ie` is an instance of `Core`, which is also shared betwen all cores and `LoadedNetwork` is just a custom class which stores the 3 return variables.

In the code I have labeled with comments 4 lines of code. In the place of these comments I tried adding scoped mutexes (something like `std::lock_guard`). I found that unless I placed a mutex in place of `// 1` I would get errors ("double free or corruption" most often). As a side note, I found that after the line `network = savedNetwork;`, `network.getFunction().get()` would be equal to `savedNetwork.getFunction().get()`, which I thought was the cause of the problem.

I think I have found a solution to this problem now. I found that it is possible to clone an ngraph function and rewrote the piece of code like this: [https://pastebin.com/k55b0B6Q](https://pastebin.com/k55b0B6Q). This gave a significant improvement, and it seems that there is no need for mutexes anymore. There remains, however, a bottleneck of `ie.LoadNetwork()`, but probably not much can be done about this?
